torchmultimodal/modules/optimizers/anyprecision.py

```diff
@@ -0,0 +1,199 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+
+from typing import Any, Dict, Iterable, Tuple, Union
+
+import torch
+from torch.optim.optimizer import Optimizer
+from typing_extensions import TypeAlias
+
+params_t: TypeAlias = Union[Iterable[torch.Tensor], Iterable[Dict[str, Any]]]
+
+
+class AnyPrecisionAdamW(Optimizer):
+    def __init__(
+        self,
+        params: params_t,
+        lr: float = 1e-3,
+        betas: Tuple[float, float] = (0.9, 0.999),
+        eps: float = 1e-8,
+        weight_decay: float = 0.0,
+        use_kahan_summation: bool = False,
+        momentum_dtype: torch.dtype = torch.float32,
+        variance_dtype: torch.dtype = torch.bfloat16,
+        compensation_buffer_dtype: torch.dtype = torch.bfloat16,
+    ) -> None:
+        """
+        AnyPrecisionAdamW: a flexible precision AdamW optimizer
+        with optional Kahan summation for high precision weight updates.
+        Allows direct control over momentum, variance and auxiliary compensation
+        buffer dtypes.
+        Optional Kahan summation is used to offset precision reduction for
+        the weight updates. This allows full training in BFloat16 (can be equal or
+        better than FP32 results in many cases) due to high precision weight upates.
+
+        This optimizer is the same AnyPrecision that was previously residing
```

Comment:
nit: I don't think this part of the doc is really needed, plus we don't wanna advertise torchdistX much.

Reply:
Ah sorry this change is actually based on a [previous comment](https://github.com/facebookresearch/multimodal/pull/477#discussion_r1349370152) of mine 😅 . Personally I am fine either way here

Reply:
I'd prefer to leave this comment mostly to avoid confusion (i.e. if someone finds there are two AnyPrecision within the larger PT repo or external references ala HF to the TorchDistX AnyPrecision).